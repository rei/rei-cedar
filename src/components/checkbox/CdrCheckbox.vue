<script setup lang="ts">
import { useCssModule, computed } from 'vue';
import type { Directive } from 'vue';
import CdrLabelWrapper from '../labelWrapper/CdrLabelWrapper.vue';
import sizeProps from '../../props/size';
import propValidator from '../../utils/propValidator';
import backgroundProps from '../../props/background';

/** Allows selecting one or more items from a list */
defineOptions({
  name: 'CdrCheckbox',
  inheritAttrs: false,
  customOptions: {},
});

const props = defineProps({
/**
     * Passes a CSS class to the label for custom styles
     */
     labelClass: String,
    /**
     * Passes a CSS class to the input for custom styles
     */
    inputClass: String,
    /**
     * Passes a CSS class to the slot wrapper for custom styles
     */
    contentClass: String,
    /**
     * Show checkbox in indeterminate state. (NOTE: this is a visual-only state and there is no logic for when to show it)
     */
    indeterminate: {
      type: [Boolean, String],
      default: false,
    },
    /** Disables the checkbox */
    disabled: Boolean,
    /**
     * The value when checked.
     */
    trueValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      default: true,
    },
    /**
     * The value when unchecked.
     */
    falseValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      default: false,
    },
    /**
     * The value when used in a checkbox group. Replaces `trueValue` and `falseValue`.
     */
    customValue: [String, Number, Boolean, Object, Array, Symbol, Function],
    /**
     * Sets the background color the input is rendered on
     * @values primary, secondary
     */
    background: backgroundProps,
    /**
     * @demoSelectMultiple false
     * @values small, medium, large
    */
    size: sizeProps,
    /**
     * Use `hide-figure` to hide the checkbox, which leaves the text label as the clickable element.
     * Add appropriate custom styles to convey selected and unselected states.
     * @values hide-figure
     */
    modifier: {
      type: String,
      default: '',
      validator: (value: string) => propValidator(value, ['', 'hide-figure']),
    },
    /** @ignore */
    modelValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
    },  
});
const emits = defineEmits({
      /**
     * Event emitted by v-model on the <input> element
     * @param modelValue
     */
     'update:modelValue': null,
});

const vIndeterminate: Directive<HTMLElement> = {
  mounted(el, binding) {
    if (binding.value) {
      el.setAttribute('indeterminate', binding.value);
      return;
    }
    el.removeAttribute('indeterminate');
  },
  updated(el, binding) {
    if (binding.value) {
      el.setAttribute('indeterminate', binding.value);
      return;
    }
    el.removeAttribute('indeterminate');
  },
};
const style = useCssModule();
const baseClass = 'cdr-checkbox';

const checkboxModel = computed({
  get() {
    return props.modelValue;
  },
  set(newValue) {
    emits('update:modelValue', newValue);
  },
});

</script>

<template>
  <cdr-label-wrapper
    :class="style[baseClass]"
    :size="size"
    :modifier="modifier"
    :label-class="labelClass"
    :content-class="contentClass"
    :background="background"
    :disabled="disabled"
  >
    <template #input>
      <input
        :class="[style['cdr-checkbox__input'], inputClass]"
        :disabled="disabled"
        type="checkbox"
        v-bind="$attrs"
        :true-value="customValue ? null : trueValue"
        :false-value="customValue ? null : falseValue"
        :value="customValue"
        v-indeterminate="indeterminate"
        v-model.lazy="checkboxModel"
      >
    </template>
    <template #svgs>
      <div :class="style['cdr-checkbox__svg-box']">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
        >
          <path
            role="presentation"
            d="M9.673 18.669h.001L19.766 6.644a1 1 0 10-1.532-1.286l-9.3
            11.085-3.169-3.776a1 1 0 10-1.532 1.286l3.875 4.618a.999.999 0 001.565.099z"
          />
        </svg>
      </div>
    </template>
    <!-- @slot Readable text for the label element -->
    <slot />
  </cdr-label-wrapper>
</template>

<style lang="scss" module src="./styles/CdrCheckbox.module.scss">
</style>
