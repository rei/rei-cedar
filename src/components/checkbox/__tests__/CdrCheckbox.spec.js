import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrCheckbox from '../CdrCheckbox.vue';

describe('CdrCheckbox', () => {
  describe('default configuration', ()=>{
    let wrapper;
    beforeEach(()=>{
      wrapper = mount(CdrCheckbox);
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
    
    it('is type checkbox', () => {
      expect(wrapper.find('input').attributes('type')).toBe('checkbox');
    });
  
    it('renders a label element', () => {
      expect(wrapper.find('label').exists()).toBe(true);
    });
  });

  describe('with a custom label class', ()=>{
    let wrapper;
    beforeEach(()=>{
      wrapper = mount(CdrCheckbox, {
        propsData: {
          labelClass: 'custom-label-class',
        },
        slots: {
          default: 'Label Test',
        },
      });
    });
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('adds a custom label class correctly', () => {
      expect(wrapper.find('.custom-label-class').exists()).toBe(true);
    });
  })

  describe('with a custom input class', ()=>{
    let wrapper;
    beforeEach(()=>{
      wrapper = mount(CdrCheckbox, {
        propsData: {
          inputClass: 'custom-input-class',
        },
        slots: {
          default: 'Label Test',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('adds a custom input class correctly', () => {
      expect(wrapper.find('.custom-input-class').exists()).toBe(true);
    });
  })

  describe('with a custom content class', ()=>{
    let wrapper;
    beforeEach(()=>{
      wrapper = mount(CdrCheckbox, {
        propsData: {
          contentClass: 'custom-content-class',
        },
        slots: {
          default: 'Label Test',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('adds a custom content class correctly', () => {
      expect(wrapper.find('.custom-content-class').exists()).toBe(true);
    });
  })

  describe('with an indeterminate prop', ()=>{
    let wrapper;
    beforeEach(async ()=>{
      wrapper = mount(CdrCheckbox, {
        propsData: {
          indeterminate: true,
        },
        slots: {
          default: 'Label Test',
        },
      });
    });

    it('renders correctly', async () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  })

  describe('when modelValue is set to false', ()=>{
    let wrapper;
    let checkbox;
    beforeEach(()=>{
      wrapper = mount(CdrCheckbox, {
        propsData: {
          modelValue: false,
        },
      });
      checkbox = wrapper.find('input');
    });

    it('a change event will cause the update:modelValue event to be emitted', () =>{
      checkbox.element.checked = true;
      checkbox.trigger('change');
      expect(wrapper.emitted()['update:modelValue'][0][0]).toBe(true);

      checkbox.element.checked = false;
      checkbox.trigger('change');
      expect(wrapper.emitted()['update:modelValue'][1][0]).toBe(false);
    })
  })

// TODO: these 2 tests no longer work properly, they are also sort of testing vue internals.
// Not sure how best to handle these...
  it('emits change events with correct values for custom checkbox', () => {
    const wrapper = mount(CdrCheckbox, {
      propsData: {
        trueValue: 'checked',
        falseValue: 'unchecked',
        modelValue: '',
      },
    });
    const cb = wrapper.find('input');
    cb.element.checked = true;
    cb.trigger('change');
    expect(wrapper.emitted()['update:modelValue'][0][0]).toBe('checked');
    cb.element.checked = false;
    cb.trigger('change');
    expect(wrapper.emitted()['update:modelValue'][1][0]).toBe('unchecked');
  });
  //
  it('emits change events with correct values for group checkbox', async () => {
    const wrapper = mount(CdrCheckbox, {
      propsData: {
        customValue: 'b',
        modelValue: ['a'],
      },
    });
    const cb = wrapper.find('input');
    cb.element.checked = true;
    cb.trigger('change');
    expect(wrapper.emitted()['update:modelValue'][0][0]).toEqual(expect.arrayContaining(['a', 'b']));
    cb.element.checked = false;
    cb.trigger('change');
    //this doesn't seem to work because the ['a', 'b'] value doesn't seem to persist
    //expect(wrapper.emitted()['update:modelValue'][1][0]).toEqual(['a']);
  });
});
