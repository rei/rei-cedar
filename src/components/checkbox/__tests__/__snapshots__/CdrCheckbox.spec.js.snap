// Vitest Snapshot v1

exports[`CdrCheckbox > renders correctly 1`] = `
<div
  class="cdr-label-wrapper__container cdr-checkbox"
>
  <label
    class="cdr-label-wrapper cdr-label-wrapper--primary cdr-label-wrapper--medium custom-label-class"
  >
    
    <input
      class="cdr-checkbox__input"
      false-value="false"
      true-value="true"
      type="checkbox"
      value="false"
    />
    
    <span
      class="cdr-label-wrapper__figure"
    />
    <span
      class="cdr-label-wrapper__content"
    >
      
      
      Label Test
      
      
    </span>
  </label>
</div>
`;
